kernel/kernel-default/src/main/java/io/delta/kernel/parquet/ArrayConverter.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (2023) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.delta.kernel.parquet;
+
+import static io.delta.kernel.parquet.ParquetConverters.initNullabilityVector;
+import static io.delta.kernel.parquet.ParquetConverters.setNullabilityToTrue;
+import java.util.Arrays;
+import java.util.Optional;
+import org.apache.parquet.io.api.Converter;
+import org.apache.parquet.io.api.GroupConverter;
+import org.apache.parquet.schema.GroupType;
+
+import io.delta.kernel.data.ColumnVector;
+import io.delta.kernel.data.vector.DefaultArrayVector;
+import io.delta.kernel.types.ArrayType;
+
+class ArrayConverter
+    extends GroupConverter
+    implements ParquetConverters.BaseConverter
+{
+    private final ArrayType typeFromClient;
+    private final ArrayCollector converter;
+
+    // working state
+    private int currentRowIndex;
+    private boolean[] nullability;
+    private int[] offsets;
+    private int collectorIndexAtStart;
+
+    public ArrayConverter(
+        int maxBatchSize,
+        ArrayType typeFromClient,
+        GroupType typeFromFile)
+    {
+        this.typeFromClient = typeFromClient;
+        final GroupType innerElementType = (GroupType) typeFromFile.getType("list");
+        this.converter = new ArrayCollector(
+            maxBatchSize,
+            typeFromClient,
+            innerElementType
+        );
+
+        // initialize working state
+        this.nullability = initNullabilityVector(maxBatchSize);
+        this.offsets = new int[maxBatchSize + 1];
+    }
+
+    @Override
+    public Converter getConverter(int fieldIndex)
+    {
+        switch (fieldIndex) {
+            case 0:
+                return converter;
+            default:
+                throw new IllegalArgumentException(
+                    "Invalid field index for a map column: " + fieldIndex);
+        }
+    }
+
+    @Override
+    public void start()
+    {
+        collectorIndexAtStart = converter.currentEntryIndex;
+    }
+
+    @Override
+    public void end()
+    {
+        int collectorIndexAtEnd = converter.currentEntryIndex;
+        this.nullability[currentRowIndex] = collectorIndexAtEnd == collectorIndexAtStart;
```

Comment:
I think this is wrong for empty arrays (I tried it out and they're read as a null value). Same thing for all the other group converters.

If start/end is called at all shouldn't nullability[currentRowIndex] = false?

Reply:
Nice catch. This could be a problem for empty array. Fixing it.

Reply:
Debugging for the empty array case and null array case, unfortunately Parquet calls GroupConvert.start and end in both cases. Looking at the Parquet-Mr provided implementation of same for [Avro conversion](https://github.com/apache/parquet-mr/blob/master/parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java#L825), they seem to have the same issue. If there is no element set call, then it is considered a null.

I added a test case for now. Once this PR is merged, will create an issue to if the behavior in other engines and if it is going to be a problem.